react-components:Add log management UI

matiasbavera opened the pull request:
## What's new

### React Components
* Added log table component customize to show different colors depending on the Log Level.
* Added a log search form.
* * Added a log management component.
* Added stories for the `SearchForm`, the `LogTable`, and `LogManagement` component.


![image](https://user-images.githubusercontent.com/11761240/106800629-75c8d180-663f-11eb-95c5-46bd25760831.png)



## Self-checks

- [X] I'm familiar with and follow this [ Typescript guideline](https://basarat.gitbook.io/typescript/styleguide)
- [X] I added unit-tests for new components
- [ ] I tried testing edge cases
- [ ] I tested the behavior of the components that interact with the backend, with an e2e test

## Discussion
* I added a pagination component to handle the logs, but if the pagination it's counter-intuitive we can change it to an infinite-scroll when we handle the pagination in the backend.

koonpeng commented:
https://kubernetes.io/docs/concepts/cluster-administration/logging/ This should give us some idea of the logging architecture we should aim for.

matiasbavera commented:
![image](https://user-images.githubusercontent.com/11761240/105092543-218af280-5a80-11eb-8ae4-2fa3eac1394f.png)
![image](https://user-images.githubusercontent.com/11761240/105092600-34052c00-5a80-11eb-84e1-340349c94ff5.png)

koonpeng commented:
this looks great, we can merge this in react components first while we develop the backend to make it useful in the dashboard

matiasbavera commented:
Adding some tests now. I'll move the fluent conf to another PR

matiasbavera commented:
Hi @koonpeng , ready for the first iteration =)

koonpeng commented on packages/react-components/stories/log-table.stories.tsx:
```diff
@@ -0,0 +1,46 @@
+import { Story, Meta } from '@storybook/react';
```
![image](https://user-images.githubusercontent.com/45189696/105649600-cd529900-5eeb-11eb-82b7-d6ef41308883.png)

![image](https://user-images.githubusercontent.com/45189696/105649612-de030f00-5eeb-11eb-809d-90b8b17d8a62.png)

The column size is changing across pages, there is also too much space given for "Level" sometimes. Having a long message in one row also causes all the rows to take more spacing, this causes some side effects like the date being split across multiple lines.

matiasbavera replied:
Fixed. 
![same width](https://user-images.githubusercontent.com/11761240/105899297-42f85b00-5ff9-11eb-87c1-761b51162d42.gif)

koonpeng commented on packages/react-components/stories/search-log-form.stories.tsx:
```diff
@@ -0,0 +1,17 @@
+import { Story, Meta } from '@storybook/react';
```
Should we replace/combine this with the log management story in dashboard?

matiasbavera replied:
Makes, sense, thanks for the feedback.

matiasbavera replied:
![image](https://user-images.githubusercontent.com/11761240/105785441-a16dfc80-5f59-11eb-84b9-f2dfa04a6320.png)

matiasbavera commented on packages/react-components/lib/date-time-picker.tsx:
```diff
@@ -0,0 +1,47 @@
+import React from 'react';
```
Yeah, I thought a native solution will be better for the browser support, but I guess I was wrong. =0

matiasbavera replied:
Had this error with Moment lib, fixed by downgrading `date-io` to 1.3.13 

```
utils.getYearText is not a function
TypeError: utils.getYearText is not a function
    at DateTimePickerToolbar (http://localhost:6006/vendors~main.e6775aeedff598925967.bundle.js:269584:18)
    at renderWithHooks (http://localhost:6006/vendors~main.e6775aeedff598925967.bundle.js:220318:18)
    at mountIndeterminateComponent (http://localhost:6006/vendors~main.e6775aeedff598925967.bundle.js:222997:13)
    at beginWork (http://localhost:6006/vendors~main.e6775aeedff598925967.bundle.js:224111:16)
    at HTMLUnknownElement.callCallback (http://localhost:6006/vendors~main.e6775aeedff598925967.bundle.js:205703:14)
    at Object.invokeGuardedCallbackDev (http://localhost:6006/vendors~main.e6775aeedff598925967.bundle.js:205752:16)
    at invokeGuardedCallback (http://localhost:6006/vendors~main.e6775aeedff598925967.bundle.js:205807:31)
    at beginWork$1 (http://localhost:6006/vendors~main.e6775aeedff598925967.bundle.js:228718:7)
    at performUnitOfWork (http://localhost:6006/vendors~main.e6775aeedff598925967.bundle.js:227672:12)
    at workLoopSync (http://localhost:6006/vendors~main.e6775aeedff598925967.bundle.js:227645:22)
```

It's a known bug. 
https://github.com/mui-org/material-ui-pickers/issues/1442


Having this warning on running test:
`
 Deprecation warning: value provided is not in a recognized RFC2822 or ISO format. moment construction falls back to js Date(), which is not reliable across all browsers and versions. Non RFC2822/ISO date formats are discouraged. Please refer to http://momentjs.com/guides/#/warnings/js-date/ for more info.`

This is on purpose because I'm removing the timezone to make those tests.

koonpeng replied:
>  This is on purpose because I'm removing the timezone to make those tests.

Can' we test with timezones? Is there a way to tell jsdom what locale settings to use?

matiasbavera replied:
Fixed moment warning by moving the date to ISO format.

matiasbavera commented on packages/dashboard/src/components/log-management.tsx:
```diff
@@ -0,0 +1,48 @@
+import React from 'react';
```
Sure. I'll make sure this component can receive the callbacks from its parent then.

matiasbavera replied:
Done

matiasbavera commented on packages/react-components/lib/log-management/log-table.tsx:
```diff
@@ -0,0 +1,103 @@
+    color: theme.palette.error.main,
+  },
+  debug: {
+    color: theme.palette.success.main,
```
Well, I thought that generally, the `bug` icons are green, so I placed green as a `debug` color, but I can switch it to theme.palette.secondary.dark, or to a darker green. Let me know what you think.

matiasbavera commented on packages/react-components/lib/log-management/search-log-form.tsx:
```diff
@@ -0,0 +1,179 @@
+
+  const searchQuery = () => {
+    if (!isFormValid()) return;
+    console.log(searchText, logLabel, logLevel, rowsCount, toLogDate, fromLogDate);
```
Yeah, Thanks.

matiasbavera commented on packages/react-components/lib/tables/table-footer-pagination.tsx:
```diff
@@ -0,0 +1,43 @@
+import React from 'react';
```
Mmm.. sorry but maybe I'm missing something, I'm using the `TablePagination` from material UI https://next.material-ui.com/components/pagination/#table-pagination . I just wrapped it into a custom component for two things: 
 * In case we have another table that needs pagination.
 * In case we need to change the paginations of all tables later (because as you said it's in `lab` version).

If none of those reasons convince you, I can put it without problem directly in `log-table.tsx` =)

matiasbavera commented on packages/react-components/stories/log-table.stories.tsx:
```diff
@@ -0,0 +1,46 @@
+    npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
+    
+    npm ERR! A complete log of this run can be found in:
+    npm ERR!     /home/ekumen/.npm/_logs/2021-01-18T21_20_07_480Z-debug.log`,
```
Actually, I don't know. But now I "fixed" that. But we can go back to that later if it's better.
![same column width](https://user-images.githubusercontent.com/11761240/105888581-e6db0a00-5feb-11eb-82ce-ef7629d151df.gif)

matiasbavera commented on packages/dashboard/src/stories/log-management.stories.tsx:
```diff
@@ -0,0 +1,11 @@
+import React from 'react';
```
![sticky](https://user-images.githubusercontent.com/11761240/105896226-4ab60080-5ff5-11eb-9d25-54fea61f0eda.gif)

matiasbavera replied:
Resolving this

matiasbavera commented:
Rebased

koonpeng commented:
Can you add some logs to the log management story so we can test the search/filter function? For the search component, do you see any use besides using it with the log management? If not then I suppose we don't really need a story for it as the log management story already covers it.

nit:
The possible of the columns sometimes change slightly when changing pages

https://user-images.githubusercontent.com/45189696/106224706-c646c600-621e-11eb-8669-0e2b4d30495e.mp4

koonpeng commented on packages/react-components/package.json:
```diff
@@ -17,8 +17,11 @@
   "author": "koonpeng@openrobotics.org",
   "license": "Apache-2.0",
   "dependencies": {
+    "@date-io/moment": "^1.3.13",
```
Which version was it that caused the problem? npm may install newer versions based on semver.

matiasbavera replied:
Well, that's weird my npm installed 2.9.2 as the last version =O. I'll not trust npm blindly anymore. We have ^2.10.6 now, that it's the current last version. It's working correctly now after some date format config

matiasbavera replied:
It's not working again, it explodes in some situations. I'll rollback to ^1.3.13. The creators of material-ui-picker also recommend using 1.x.x.
https://material-ui-pickers.dev/getting-started/installation#peer-library

koonpeng commented on packages/react-components/tests/log-management/log-management.test.tsx:
```diff
@@ -0,0 +1,30 @@
+
+it('smoke test', async () => {
+  // Added the waitFor because this component is updating a state inside a useEffect.
+  await waitFor(() => {
+    render(<LogManagement getLabels={getLabelsPromise} getLogs={getLogsPromise} />);
+  });
```
Would it be better to use `act`?

matiasbavera replied:
Actually, from what I [read](https://github.com/testing-library/react-testing-library/issues/535#issuecomment-660267339), the RTL render function is already wrapped in an act by default. In this case, we need the waitFor because we need to wait until the `useEffect` async content is executed. 

https://testing-library.com/docs/dom-testing-library/api-async/#waitfor

Offtopic: during my search about the problem I found this interesting post https://kentcdodds.com/blog/common-mistakes-with-react-testing-library =)

koonpeng replied:
thanks for the explaination

matiasbavera commented:
> Can you add some logs to the log management story so we can test the search/filter function? For the search component, do you see any use besides using it with the log management? If not then I suppose we don't really need a story for it as the log management story already covers it.

Oh sorry, now I get it, I thought that the function of that word search engine was just to send it with the request and that the backend filter the data based on that.

As an option for what I think you are saying we could implement this: https://material-ui.com/components/data-grid/
To search within the data loaded in the UI

Let me know what you think

matiasbavera commented:
> nit:
> The possible of the columns sometimes change slightly when changing pages
> 
>  Peek.2021-01-29.10-41.mp4

Thanks for the feedback, fixed, the label and the content are not moving horizontally anymore. That movement you see is vertical because of the vertical space that the following message use

```
  npm ERR! code ELIFECYCLE
    npm ERR! errno 1
    npm ERR! react-components@0.0.1 build: 'npm run lint && tsc --build'
    npm ERR! Exit status 1
    npm ERR! 
    npm ERR! Failed at the react-components@0.0.1 build script.
    npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
    
    npm ERR! A complete log of this run can be found in:
    npm ERR!     /home/ekumen/.npm/_logs/2021-01-18T21_20_07_480Z-debug.log 
```

![fixed table](https://user-images.githubusercontent.com/11761240/106319128-b72c3c00-624f-11eb-8adb-8802609b4d71.gif)


Also, I'm using rem now as the unit metric, for the table to scale correctly on zoom.

koonpeng commented:
> > Can you add some logs to the log management story so we can test the search/filter function? For the search component, do you see any use besides using it with the log management? If not then I suppose we don't really need a story for it as the log management story already covers it.
> 
> Oh sorry, now I get it, I thought that the function of that word search engine was just to send it with the request and that the backend filter the data based on that.
> 
> As an option for what I think you are saying we could implement this: https://material-ui.com/components/data-grid/
> To search within the data loaded in the UI
> 
> Let me know what you think

I haven't thought about how the searching should work but I guess a mix of server queries and client filtering would be best. We should at least be able to query the logs based on time so that we don't download the entire history. More specific filters or sorting can be done on the client to be more responsive.

The data grid looks really nice, is it fully client side or does it support dynamicly querying the server?

koonpeng reviewed (commented):
I did not think about how the search form works, I instinctly thought the filters are done client side, which way do you think is better? Fully server side or a mix of server and client filtering?

Suppose we go with fully server side, then I guess it makes sense to separate the form and the viewer. I think it would still be good to have a mock data source so we can easily test out the search & filter capabilities, but that is probably a problem we need to deal with later when we do the implementation in `dashboard`.

In that case, I suppose there is no use for the log management story as it isn't useful without a backend, it makes more sense to move it back into `dashboard`. But if we do mixed server/client search & filter, then it makes sense for log management to be in react-components, it will take in an array of logs instead of a query function and we can test it in a story.

matiasbavera commented:
> I did not think about how the search form works, I instinctly thought the filters are done client side, which way do you think is better? Fully server side or a mix of server and client filtering?

Yeah, I think a mix should be good. So, as you said, the main idea is to handle the heavy query in the backend, e.g. search between dates, labels (optional), and log type(optional) (dashboard, rmf-core), etc. The two things that drive me to do this is that:
* Having these pre-filters we will reduce the amount of data that we send to the front-end
* By handling less data, it will be easier for the browser to render and handle the logs in memory. 

So, when we have the response we can filter it in the UI, to make things easier for the final user.


I decided to drop this option for all the limitations that the community version has (before it did not have so many limitations) 
https://material-ui.com/components/data-grid/getting-started/#feature-comparison


And I choose https://material-table.com/#/docs/get-started which is pretty cool and configurable.

![table](https://user-images.githubusercontent.com/11761240/106520646-daf3ba00-64bb-11eb-8f99-fa5fe1c86e84.gif)


So I think the flow will be something like this. The search button will request data to the backend, and the table will show the response and it'll provide tools to filter and sort the data:
![flow](https://user-images.githubusercontent.com/11761240/106520886-332abc00-64bc-11eb-9c80-3b92531a8254.gif)


Let me know what you think.

koonpeng commented:
I think the idea is right, some nits about the design

1. It would be nice if the log level filter has a selectable list.
2. It's not obvious to me how the time filter works, does it return results close to the time, before, or after it?
3. It's not obvious what the "search" field in the table is searching for. Is it a combination of log and message filter?
4. It can be confusing to have 2 "search" field/button. Some of which like log level and message text is duplicated, in general I think its more clear if we remove some of the existing server query params. We can also try renaming the server search button to something like "retrieve" to avoid confusion.

matiasbavera commented:
> I think the idea is right, some nits about the design

Awesome!

> 1. It would be nice if the log level filter has a selectable list.

Yeah, makes sense. Done!

> 2. It's not obvious to me how the time filter works, does it return results close to the time, before, or after it?

It's the exact date, I'll find a workaround to filter between datetimes. 

> 3. It's not obvious what the "search" field in the table is searching for. Is it a combination of log and message filter?
> 4. It can be confusing to have 2 "search" field/button. Some of which like log level and message text is duplicated, in general I think its more clear if we remove some of the existing server query params. We can also try renaming the server search button to something like "retrieve" to avoid confusion.

Yeah, you're right, to avoid confusion:
* Changed `Search` -> `Retrieve Logs`
* Removed the search input component
* Removed rows count

koonpeng commented:
> It's the exact date, I'll find a workaround to filter between datetimes.

How about leaving the date filtering to the server side query? By default it will retrieve the last X days, then the user can use the "retrieve logs" form to change the date range. We do lose the function of near instant time filtering but I think it is a good compromise, maybe if we really want it we can do a "smart query" where we only retrieve the logs which have not been already retrieved, if the date range falls into a previously retrieved set then we just reuse it. But that looks like a non-trivial amount of logic, we have to take care of edge cases like "refreshing" the latest logs.

matiasbavera commented:
> How about leaving the date filtering to the server side query? By default it will retrieve the last X days, then the user can use the "retrieve logs" form to change the date range. We do lose the function of near instant time filtering but I think it is a good compromise, maybe if we really want it we can do a "smart query" where we only retrieve the logs which have not been already retrieved, if the date range falls into a previously retrieved set then we just reuse it. But that looks like a non-trivial amount of logic, we have to take care of edge cases like "refreshing" the latest logs.

Yeah, makes sense. We can do that. But I'd change the word `days` per `hours` because of the number of logs. Ok, I'll remove the date filter from the table and we can manage the default load from the component in the dashboard, which it'll be created when we have the logs (currently working on a PR with fluentD, I'm getting logs from containers, log files and HTTP protocol and exporting to plain text files ) and backend for it.

koonpeng reviewed (approved):
thanks for this, just a couple of nits

1. did the font changed?
2. We can reduce the width of the "Log Level" column.
3. Sort the log levels according to the severity
![image](https://user-images.githubusercontent.com/45189696/106832178-28467600-66cc-11eb-9bad-c2d64a1153f5.png)
4. Show all the log levels without scrolling, reducing the padding/margin between the lines might help.
5. ~Would it be more intuitive if all the checkboxes are checked by default? (I think it shows all logs by default)~. It seems like it is the default behavior for MaterialTable, I guess this is more inline with how office apps works.

matiasbavera commented:
> thanks for this, just a couple of nits
> 
> 1. did the font changed?
> 2. We can reduce the width of the "Log Level" column.
> 3. Sort the log levels according to the severity
>    ![image](https://user-images.githubusercontent.com/45189696/106832178-28467600-66cc-11eb-9bad-c2d64a1153f5.png)
> 4. Show all the log levels without scrolling, reducing the padding/margin between the lines might help.
> 5. ~Would it be more intuitive if all the checkboxes are checked by default? (I think it shows all logs by default)~. It seems like it is the default behavior for MaterialTable, I guess this is more inline with how office apps works.

1. Yeah,  you're right, good catch, that's weird because it's built on top of material-ui. But I unified the font now. 
2. Done!
![image](https://user-images.githubusercontent.com/11761240/106934374-cea87080-66f8-11eb-8634-cbb4a9bc497d.png)
3. Done!
4.
